### DISCUSSION TITLE
Is it possible to map characteristic data bytes to a typed entity

### DISCUSSION THREAD

**Discussion by still-learnin** [christoph2/pyA2L#36] [Q&A] [answered]
Using pyxcp I can request a characteristic from the slave based upon its address and length. If I then want to interpret the data bytes returned as a meaningful typed value I can then look at various attributes of the characteristics class to determine how to interpret this, i.e. string, signed/unsigned value, array etc. etc. Just in case I didn't find it or realise it existed, does the pyA2L package already handle this for me (i.e. something that returns a suitably typed entity?

**Comment by christoph2** [comment, marked as answer]
I'm glad you're asking!!! 🙂

This is exactly were [asamint](https://github.com/christoph2/asamint) kicks in.
As soon as you're leaving the world of scalar characteristics, things are getting complex and complicated -- to name the (one) enemy: RECORD_LAYOUT; there are so many details to consider (some information could be easily extracted from your a2ldb, other information needs to be dynamically created with the help of a memory image), the spec. is nebulous at many points -- you certainly won't do this at home...
But I've spent a considerable amount of time to get this working (i.e. carefully comparing the converted values, curves, maps, ..., with the outcome of CANape).
Yes, with **asamint** it's possible to create ASAM CDF and Damos DCM files (other formats, like Vector .PAR and PaCo will follow). I'm also able to parse these formats; the main issue is currently to find a common representation for these quite different formats and a corresponding API.
There's some clean-up required, but I'm trying hard to present a working example  even today.

Cheers Christoph

**Comment by christoph2** [comment]
OK, in the [asamint examples directory](https://github.com/christoph2/asamint/tree/master/asamint/examples) you'll find to calibration examples:

- **create_calparam_file.{cmd,sh}**:
  Creates an .CDFX file from the official CDF20demo, A2L and HEX. Run it to check if everything is installed correctly.

- **cdf_from_xcp.{cmd,sh}**: This is your starting point when it comes to parameter extraction and may not work out-of-the-box. 

asamint uses two configuration files (parameters are mostly self-explanatory): 

- project specific (.apr), may also contain every pyXCP parameter.
- experiment specific (.exp)

**Comment by christoph2** [comment]
I've forgot to mention an interim solution:
objutils contains an API for typified access of HEX files (used internally by **asamint**).
s. the [README](https://github.com/christoph2/objutils/)

and the [source](https://github.com/christoph2/objutils/blob/master/objutils/image.py)
